### ISSUE TITLE
install error with ubuntu 14.04 LTS on VM with released version 0.9.9

### ISSUE THREAD

**Issue by gssrok** [BinaryAnalysisPlatform/bap#374]
I've tried to install OCaml and BAP from a source code but it is stuck.

According to this github and its wiki, I installed opam 1.2.2  and other packeges for BAP indicated by depext.
However,` libncurses5-dev is installed instead of ncurses-dev package.

Here's an error message and reports by a make command.

```
Configuration:
ocamlfind: ........................................... /home/lab1/.opam/4.02.3/bin/ocamlfind
ocamlc: .............................................. /home/lab1/.opam/4.02.3/bin/ocamlc.opt
ocamlopt: ............................................ /home/lab1/.opam/4.02.3/bin/ocamlopt.opt
ocamlbuild: .......................................... /home/lab1/.opam/4.02.3/bin/ocamlbuild
Package name: ........................................ bap
Package version: ..................................... 0.9.9
os_type: ............................................. Unix
system: .............................................. linux
architecture: ........................................ amd64
ccomp_type: .......................................... cc
ocaml_version: ....................................... 4.02.3
standard_library_default: ............................ /home/lab1/.opam/4.02.3/lib/ocaml
standard_library: .................................... /home/lab1/.opam/4.02.3/lib/ocaml
standard_runtime: .................................... /home/lab1/.opam/4.02.3/bin/ocamlrun
bytecomp_c_compiler: ................................. gcc -O -fno-defer-pop -Wall -D_FILE_OFFSET_BITS=64 -D_REENTRANT -O -fPIC
native_c_compiler: ................................... gcc -O -Wall -D_FILE_OFFSET_BITS=64 -D_REENTRANT
model: ............................................... default
ext_obj: ............................................. .o
ext_asm: ............................................. .s
ext_lib: ............................................. .a
ext_dll: ............................................. .so
default_executable_name: ............................. a.out
systhread_supported: ................................. true
Install architecture-independent files dir: .......... /usr/local
Install architecture-dependent files in dir: ......... $prefix
User executables: .................................... $exec_prefix/bin
System admin executables: ............................ $exec_prefix/sbin
Program executables: ................................. $exec_prefix/libexec
Read-only single-machine data: ....................... $prefix/etc
Modifiable architecture-independent data: ............ $prefix/com
Modifiable single-machine data: ...................... $prefix/var
Object code libraries: ............................... $exec_prefix/lib
Read-only arch-independent data root: ................ $prefix/share
Read-only architecture-independent data: ............. $datarootdir
Info documentation: .................................. $datarootdir/info
Locale-dependent data: ............................... $datarootdir/locale
Man documentation: ................................... $datarootdir/man
Documentation root: .................................. $datarootdir/doc/$pkg_name
HTML documentation: .................................. $docdir
DVI documentation: ................................... $docdir
PDF documentation: ................................... $docdir
PS documentation: .................................... $docdir
findlib_version: ..................................... 1.6.2
is_native: ........................................... true
suffix_program: ...................................... 
Remove a file.: ...................................... rm -f
Remove a directory.: ................................. rm -rf
Turn ocaml debug flag on: ............................ true
Turn ocaml profile flag on: .......................... false
Compiler support generation of .cmxs.: ............... true
OCamlbuild additional flags: ......................... 
piqic_path: .......................................... /home/lab1/.opam/4.02.3/bin/piqic-ocaml
piqic_flags: ......................................... 
piqic: ............................................... /home/lab1/.opam/4.02.3/bin/piqic-ocaml
c++ compiler: ........................................ /usr/bin/c++
C compiler optimization level: ....................... -O2
llvm version (e.g., 3.4): ............................ 3.4
llvm-config executable: .............................. /usr/bin/llvm-config-3.4
Build serialization library: ......................... true
Links with llvm in a static mode: .................... true
Build and run benchmarks: ............................ false
Build C++ interface to dissassemblers backend: ....... true
Build BAP server: .................................... true
Build ZMQ transports for BAP server: ................. false
Build with llvm backend: ............................. true
Create documentations: ............................... true
Compile tests executable and library and run them: ... false
camlp4o: ............................................. /home/lab1/.opam/4.02.3/bin/camlp4o
pkg_bin_prot_syntax: ................................. /home/lab1/.opam/4.02.3/lib/bin_prot
pkg_camlp4: .......................................... /home/lab1/.opam/4.02.3/lib/ocaml/camlp4
pkg_comparelib_syntax: ............................... /home/lab1/.opam/4.02.3/lib/comparelib
pkg_core_kernel: ..................................... /home/lab1/.opam/4.02.3/lib/core_kernel
pkg_enumerate_syntax: ................................ /home/lab1/.opam/4.02.3/lib/enumerate
pkg_faillib_syntax: .................................. /home/lab1/.opam/4.02.3/lib/faillib
pkg_fieldslib_syntax: ................................ /home/lab1/.opam/4.02.3/lib/fieldslib
pkg_herelib_syntax: .................................. /home/lab1/.opam/4.02.3/lib/herelib
pkg_pa_ounit_syntax: ................................. /home/lab1/.opam/4.02.3/lib/pa_ounit
pkg_sexplib_syntax: .................................. /home/lab1/.opam/4.02.3/lib/sexplib
pkg_variantslib_syntax: .............................. /home/lab1/.opam/4.02.3/lib/variantslib
pkg_dynlink: ......................................... /home/lab1/.opam/4.02.3/lib/ocaml
pkg_fileutils: ....................................... /home/lab1/.opam/4.02.3/lib/fileutils
pkg_findlib: ......................................... /home/lab1/.opam/4.02.3/lib/findlib
pkg_zarith: .......................................... /home/lab1/.opam/4.02.3/lib/zarith
pkg_uuidm: ........................................... /home/lab1/.opam/4.02.3/lib/uuidm
pkg_ocamlgraph: ...................................... /home/lab1/.opam/4.02.3/lib/ocamlgraph
pkg_bitstring: ....................................... /home/lab1/.opam/4.02.3/lib/bitstring
pkg_bitstring_syntax: ................................ /home/lab1/.opam/4.02.3/lib/bitstring
pkg_camlzip: ......................................... /home/lab1/.opam/4.02.3/lib/camlzip/../zip
piqi: ................................................ /home/lab1/.opam/4.02.3/bin/piqi
pkg_piqirun: ......................................... /home/lab1/.opam/4.02.3/lib/piqirun
pkg_piqirun_pb: ...................................... /home/lab1/.opam/4.02.3/lib/piqirun
pkg_piqirun_ext: ..................................... /home/lab1/.opam/4.02.3/lib/piqirun
pkg_compiler_libs: ................................... /home/lab1/.opam/4.02.3/lib/ocaml/compiler-libs
pkg_lwt: ............................................. /home/lab1/.opam/4.02.3/lib/lwt
pkg_lwt_unix: ........................................ /home/lab1/.opam/4.02.3/lib/lwt
pkg_lwt_log: ......................................... /home/lab1/.opam/4.02.3/lib/lwt
pkg_ezjsonm: ......................................... /home/lab1/.opam/4.02.3/lib/ezjsonm
pkg_uri: ............................................. /home/lab1/.opam/4.02.3/lib/uri
pkg_cohttp_lwt: ...................................... /home/lab1/.opam/4.02.3/lib/cohttp
pkg_re: .............................................. /home/lab1/.opam/4.02.3/lib/re
pkg_cmdliner: ........................................ /home/lab1/.opam/4.02.3/lib/cmdliner
pkg_re_posix: ........................................ /home/lab1/.opam/4.02.3/lib/re
pkg_curl: ............................................ /home/lab1/.opam/4.02.3/lib/curl
pkg_ocamlbuild: ...................................... /home/lab1/.opam/4.02.3/lib/ocaml/ocamlbuild

ocaml setup.ml -quiet -build 
Finished, 1 target (0 cached) in 00:00:01.
File "_tags", line 949, characters 36-44:
Warning: the tag "pkg_core" is not used in any flag or dependency declaration, so it will have no effect; it may be a typo. Otherwise you can use `mark_tag_used` in your myocamlbuild.ml to disable this warning.
File "_tags", line 984, characters 25-33:
Warning: the tag "pkg_core" is not used in any flag or dependency declaration, so it will have no effect; it may be a typo. Otherwise you can use `mark_tag_used` in your myocamlbuild.ml to disable this warning.
+ /home/lab1/.opam/4.02.3/bin/ocamlfind ocamlopt -c -g -annot -bin-annot -short-paths -I lib/bap -I lib/bap_types -I lib/bap_image -I lib/bap_elf -I lib/bap_image -I lib/bap_dwarf -I lib/bap_image -I lib/bap_disasm -I lib/bap_disasm -I lib/bap_sema -I lib/bap -syntax camlp4o -package bin_prot.syntax -package bitstring -syntax camlp4o -package bitstring.syntax -package camlp4 -package camlzip -syntax camlp4o -package comparelib.syntax -package core_kernel -package dynlink -syntax camlp4o -package enumerate.syntax -syntax camlp4o -package faillib.syntax -syntax camlp4o -package fieldslib.syntax -package fileutils -package findlib -syntax camlp4o -package herelib.syntax -package ocamlgraph -syntax camlp4o -package pa_ounit.syntax -package piqirun -package piqirun.ext -package piqirun.pb -syntax camlp4o -package sexplib.syntax -package uuidm -syntax camlp4o -package variantslib.syntax -package zarith -I lib/bap_types -I lib/bap -o lib/bap_types/bap_graph_regular.cmx lib/bap_types/bap_graph_regular.ml
File "lib/bap_types/bap_graph_regular.ml", line 1:
Error: I/O error: as -o 'lib/bap_types/bap_graph_regular.o' '/tmp/camlasm19f10a.s': Cannot allocate memory
Command exited with code 2.
Compilation unsuccessful after building 164 targets (0 cached) in 00:03:07.
make: *** [build] Error 1
```

I tried to install BAP from a source code obtained by `git clone https://github.com/BinaryAnalysisPlatform/bap.git`.
However a compiler reports an error like this:

```
lab1@ubuntu:~/bap$ make
ocaml setup.ml -quiet -build 
ocamlfind ocamlopt -package unix -package ocamlbuild -linkpkg -package core_kernel -package findlib -package ppx_driver.ocamlbuild myocamlbuild.ml /home/lab1/.opam/4.02.3/lib/ocaml/ocamlbuild/ocamlbuild.cmx -o myocamlbuild
+ ocamlfind ocamlopt -package unix -package ocamlbuild -linkpkg -package core_kernel -package findlib -package ppx_driver.ocamlbuild myocamlbuild.ml /home/lab1/.opam/4.02.3/lib/ocaml/ocamlbuild/ocamlbuild.cmx -o myocamlbuild
ocamlfind: Package `ppx_driver.ocamlbuild' not found
Command exited with code 2.
make: *** [build] Error 1

```
This is strange because I verified a ppx_driver package is alread installed.

```
lab1@ubuntu:~/bap$ opam install ppx_driver
[NOTE] Package ppx_driver is already installed (current version is
       113.09.00).

```
Please let me know how to fix it. Thanks.

**Comment by ivg** [comment]
This are two reports in one :)

# PR 1

> Error: I/O error: as -o 'lib/bap_types/bap_graph_regular.o' '/tmp/camlasm19f10a.s': Cannot allocate memory> 

The error is `Cannot allocate memory`. You should increase you physical memory, either in you VM or by adding more memory chips. This `bap_graph_regular` is a crazy module that defines hundreds of graphs. In the master version of graph it is much smaller, as in the newer version of OCaml we don't need this anymore. So if you can't increase you physical memory, then there is always an option to switch to a newer bap.1.0.0~alpha, that soon will be released. 

# PR 2

Building current bap manually from source is a little bit complicated issue. I didn't update our wiki with new instructions yet. So it is better to use `opam` to obtain the latest version. There're two methods, adding our alpha `opam-repository` to get an access for more than thirty bap packages (this is the preferred). Or pinning bap to the github version (not preferred in general, but useful for testing purposes - takes much more time). 

## Installing alpha version from opam-repository

First you need to add our `opam-repository`. You need to do this once:

```
opam repository add bap git://github.com/BinaryAnalysisPlatform/opam-repository.git
```
This will add a new repository of packages, that contains `bap` packages, as well as all other packages. At least it is guaranteed that this is a working snapshot of packages.

Don't forget to do `opam update` to get the latest version:
```
opam update
```

And finally, install `bap` metapackage with:

```
opam install bap
```

Note: be careful, if you have too many jobs, and not enough memory you can get out of memory again.

**Comment by gssrok** [comment]
Hi. I read your comments above and tried to install BAP alpha version.
At this time, opam reported me to install ocurl 0.7.0 for BAP pined to alpha version.
After ocurl installed, I put git clone git://.... and opam install bap --deps-only command to install BAP from source and opam reported BAP pagkage is already up-to-date.

However, when I put make command, following error is reported.

make: *** No rule to make target `setup.ml', needed by `build'.  Stop.

After ./configure is done, make command results such errors.

make:
ocaml setup.ml -quiet -build 
ocamlfind ocamlopt -package unix -package ocamlbuild -linkpkg -package core_kernel -package findlib -package ppx_driver.ocamlbuild myocamlbuild.ml /home/lab1/.opam/4.02.3/lib/ocaml/ocamlbuild/ocamlbuild.cmx -o myocamlbuild
# No parallelism done

make install:
ocaml setup.ml -quiet -install 
sh tools/build_plugins.sh


Is BAP installation guide from source not available now?
Please leave comments.
Sorry for bothering you. Thank you!

**Comment by gssrok** [comment]
I installed BAP 0.9.9 from source tree with following commands:

sudo add-apt-repository --yes ppa:avsm/ppa
sudo apt-get update
sudo apt-get --yes install opam
opam init --comp=4.02.3
eval `opam config env`

opam repository add bap git://github.com/BinaryAnalysisPlatform/opam-repository.git
opam update

wget https://github.com/BinaryAnalysisPlatform/bap/archive/v0.9.9.tar.gz
tar v0.9.9.tar.gz

(in the bap 0.9.9 source directory)
opam install depext
opam install ocurl.0.7.0
opam depext bap.0.9.9
opam install bap.0.9.9 --deps-only
make

Some warinings are reported while installation:

```
ocaml setup.ml -quiet -build 
Finished, 1 target (0 cached) in 00:00:00.
File "_tags", line 949, characters 36-44:
Warning: the tag "pkg_core" is not used in any flag or dependency declaration, so it will have no effect; it may be a typo. Otherwise you can use `mark_tag_used` in your myocamlbuild.ml to disable this warning.
File "_tags", line 984, characters 25-33:
Warning: the tag "pkg_core" is not used in any flag or dependency declaration, so it will have no effect; it may be a typo. Otherwise you can use `mark_tag_used` in your myocamlbuild.ml to disable this warning.
```

Here's a list of files after an installation is done.

apt.deps                 fsi_benchmark_main.native@  preconfig.ml
apt.opt.deps             lib/                        python/
bapbuild.ml              lib_test/                   README.md
bapbuild.native@         LICENSE                     setup.data
bapdoc.ml                lwt/                        setup.log
bap_main.native@         Makefile                    setup.ml
bap_mc.native@           myocamlbuild.ml             setup.ml.in
baptop*                  myocamlbuild.ml.in          setup.py
benchmarks/              _oasis                      src/
_build/                  opam                        start_server.native@
byteweight_main.native@  opam.deps                   _tags
CHANGES.md               opam.opt.deps               _tags.in
configure*               postinstall.ml              test.sh*
doc/                     postinstall.ml.ab

I wonder this is a correct result
because bapbuild is not found under a _build/ directory.
Only an executable bapbuild.native file and myocamlbuild file are in that directory.

I'm so sorry for bothering you.
Thank you for your kind answer.

**Comment by ivg** [comment]
Tell me please, do you want just to get latest bap installed, to get bap.0.9.9, or you want to learn how to compile it from source code?


#### To get the latest `bap` 

Use the following instructions:

For the former, you need to use opam:
```
opam repository add bap git://github.com/BinaryAnalysisPlatform/opam-repository.git
opam update
opam install bap
````
This will install bap metapackage, including `bap` frontend, and a dozen of plugins.


#### To get bap.0.9.9

Just do
```
opam install bap.0.9.9
```

#### To install bap from source code

```
git clone git@github.com:BinaryAnalysisPlatform/bap.git
cd bap
./configure --prefix=`opam config var prefix` --enable-everything
make 
make install
```


Note: all instructions assume a clean opam state with bap dependencies installed.

**Comment by rvantonder** [comment]
Can we update the README with

**Comment by abarbaro** [comment]
hello,
i have installed bap from source code but when i try it I have the following  error:

File "/usr/lib/python2.7/subprocess.py", line 711, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1340, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory

I pretty sure it is because i don't have bap-server. Can someone explain me why bap-server is not installed?

Thanks a lot,
Alberto

**Comment by abarbaro** [comment]
Hi,
if I try with opam I have this error when I try to install pa_sexp_conv:

# File "syntax/pa_sexp_conv.ml", line 1630, characters 62-80:
# Error: Unbound value Gen.string_of_ctyp
# Command exited with code 2.
# Makefile:21: recipe for target 'build' failed

Any idea?

Thanks

**Comment by ivg** [comment]
`bap-server` is temporary disabled in the 1.0.0~alpha, see #380, for more information (TL;DR waiting for upstream libraries). 

> if I try with opam I have this error when I try to install pa_sexp_conv
Ah, this is a nasty new bug in one upstream library, see #412. You're unlucky to that opam decided to install it, and we're guilty that we didn't predict, that this library will break :) (actually the library didn't exist at the time of release of bap). A possible solution, would be to help the constraint solver to choose the proper library:
```
opam install 'conduit=0.10.0' bap.0.9.9
```

If this will work, please report it back, so that I can push this constraint into the upstream repository.

**Comment by abarbaro** [comment]
Hi,
I tried as you said and I have this error:

# lib/bap_disasm/libllvm_stubs.a(llvm_disasm.o): In function `bap::llvm_disassembler::create(char const*, char const*, int)':
# /home/al/.opam/4.02.2/build/bap.0.9.9/_build/lib/bap_disasm/llvm_disasm.cpp:123: undefined reference to `llvm::TargetRegistry::lookupTarget(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)'
# collect2: error: ld returned 1 exit status
# File "caml_startup", line 1:
# Error: Error during linking
# Command exited with code 2.
# Makefile:7: recipe for target 'build' failed
### stderr ###
# make: *** [build] Error 1



=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions failed
  ∗  install bap 0.9.9
No changes have been performed

=-=- bap.0.9.9 troobleshooting =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
=> This package relies on external (system) dependencies that may be missing. `opam depext bap.0.9.9' may help you find the correct installation for your system.
username@username-X750JA:~$

The problem is that the only missing dependency is ncurses-dev but I have already installed libncurses5-dev ( I'm using ubuntu )

Any idea?

Thanks

**Comment by ivg** [comment]
looks like that you have an incompatible version of llvm, currently we support only llvm-3.4, make sure that you have this version installed.

**Comment by abarbaro** [comment]
Hi,
I have installed llvm-3.4 from the source code but I have the same error... Do you need any other specific information?

Thanks

**Comment by abarbaro** [comment]
Hi,
I found a similar problem here https://github.com/JuliaLang/julia/issues/10201. I tried to apply the same solution but no luck.

**Comment by ivg** [comment]
if you have several versions installed it is hard to predict which pne would be chosen. The configuration script will look for an `llvm-config` program eith optional version suffixes. Try to run `llvm-config`, so that we can figure out what version is taken. Even if you have different versions on your machine, or if you can't install llvm into a system path, it is still possible to pass extra information to the configuration script, via the environment variable, e.g.,
```
llvm_config=<path-to-llvm-config> opam install conduit.0.10.0 bap.0.9.0
```

p.s. the newer bap-1.0.0 is much more robust, we will release it very soon. And python support is on the way, guys from conduit promised to release it in this weekend, so hopefully everything will be ready on the next week.

**Comment by abarbaro** [comment]
Hi,
no problem, I'll wait the next release this next week. Anyway, I would like to paste the new error that maybe could help to apply other fixes in the next release:

[...]
lib/bap_image/bap_llvm_binary.hpp:231:15: error: no member named 'begin_symbols' in 'llvm::object::ObjectFile'
     read(obj->begin_symbols(),
          ~~~  ^
 lib/bap_image/bap_llvm_binary.hpp:232:15: error: no member named 'end_symbols' in 'llvm::object::ObjectFile'
          obj->end_symbols(),
          ~~~  ^
 fatal error: too many errors emitted, stopping now [-ferror-limit=]
 1 warning and 20 errors generated.
 Command exited with code 1.
 Makefile:7: recipe for target 'build' failed
stderr
make: *** [build] Error 1

If not useful I'll just wait and stop pasting the errors...

Thanks

**Comment by ivg** [comment]
yep, this all indicates that the version is different. I think that on our side we should be more careful, and do not even try to compile with an incompatible version of llvm.

**Comment by abarbaro** [comment]
Hi,
sorry I forgot to reply. Actually I have temporarily solved installing bap 0.9.7. I'll wait the next official release.

Thanks,
Alberto

**Comment by ivg** [comment]
Doing our best, to make it as soon as possible!
